code/models/DocumentationPage.php

```diff
@@ -280,38 +274,77 @@ public function Link($short = false)
      */
     public function populateMetaDataFromText(&$md, $removeMetaData = false)
     {
-        if ($md) {
-            // get the text up to the first empty line
-            $extPattern = "/^(.+)\n\r*\n/Uis";
-            $matches = preg_match($extPattern, $md, $block);
-
-            if ($matches && $block[1]) {
-                $metaDataFound = false;
-
-                // find the key/value pairs
-                $lines = preg_split('/\v+/', $block[1]);
-                $key = '';
-                $value = '';
-                foreach ($lines as $line) {
-                    if (strpos($line, ':') !== false) {
-                        list($key, $value) = explode(':', $line, 2);
-                        $key = trim($key);
-                        $value = trim($value);
-                    } else {
-                        $value .= ' ' . trim($line);
-                    }
-                    if (property_exists(get_class(), $key)) {
-                        $this->$key = $value;
-                        $metaDataFound = true;
-                    }
+        if (!$md) {
+            return;
+        }
+
+        // See if there is YAML metadata block at the top of the document. e.g.
+        // ---
+        // property: value
+        // another: value
+        // ---
+        //
+        // If we found one, then we'll use a YAML parser to extract the
+        // data out and then remove the whole block from the markdown string.
+        //
+        // We expect to get 3 matches from preg_split() which match the start
+        // --- part (0), the YAML content part (1), and the end --- part (2).
+        //
+        $yamlBlockParts = preg_split('/[\n]*[-]{3}[\n]/', $md, 3);
+        if (count($yamlBlockParts) === 3) {
+            $data = Yaml::parse($yamlBlockParts[1]);
+            if (!is_array($data)) {
+                return;
+            }
+            foreach ($data as $key => $value) {
+                if (!property_exists(get_class($this), $key)) {
+                    continue;
                 }
+                $this->$key = $value;
+            }
+            // Remove any YAML metadata block that may be at the top of the file
+            // Ensure limit is set to 1 on preg_replace so we're only replacing
+            // the first occurence of the block in the file, and not potentially
+            // more stuff later down.
+            if ($removeMetaData) {
+                $md = preg_replace('/---(.|\n)*?---/', '', $md, 1);
```

Comment:
I think you need to anchor this to the file beginning. I'd also suggest to use exactly the same regex as earlier on, to ensure you don't trigger weirdly.